### DISCUSSION TITLE
sync_business_shared_folders = "true": what if anybody deletes a file from a shared library where I have read-only access?

### DISCUSSION THREAD

**Discussion by rrodrigueznt** [abraunegg/onedrive#2087] [General]
Hi!

Running `onedrive v2.4.20` on a `CentOS Stream release 8` server. I want a fully updated local read-only copy of several OneDrive accounts and SharePoint libraries. Several services are running without a glitch. Great tool! @abraunegg, thanks for your work!

When `download_only = "false"`, I can understand the following output I get from `systemctl status [service name]`: `onedrive` downloads a file from a shared library; somebody removed the file from that shared library; `onedrive` tries to upload its local copy as it detects it is lacking in the server side, and it fails as it has read-only access.

```
Aug 12 08:53:38 e220i059h057w.inv.usc.es onedrive[11480]: Initializing the Synchronization Engine ...
Aug 12 08:53:38 e220i059h057w.inv.usc.es onedrive[11480]: Initializing monitor ...
Aug 12 08:53:38 e220i059h057w.inv.usc.es onedrive[11480]: OneDrive monitor interval (seconds): 300
Aug 12 08:58:19 e220i059h057w.inv.usc.es onedrive[11480]: Uploading new file ./AUTORIZACIÓNS GASTOS_2022/7N07_Carlos Salgado/C-003299-22-04_signed.pdf ...
Aug 12 08:58:20 e220i059h057w.inv.usc.es onedrive[11480]: skipped.
Aug 12 08:58:20 e220i059h057w.inv.usc.es onedrive[11480]: ERROR: Microsoft OneDrive API returned an error with the following message:
Aug 12 08:58:20 e220i059h057w.inv.usc.es onedrive[11480]:   Error Message:    HTTP request returned status code 403 (Forbidden)
Aug 12 08:58:20 e220i059h057w.inv.usc.es onedrive[11480]:   Error Reason:     Access denied
Aug 12 08:58:20 e220i059h057w.inv.usc.es onedrive[11480]:   Error Timestamp:  2022-08-12T06:58:20
Aug 12 08:58:20 e220i059h057w.inv.usc.es onedrive[11480]:   API Request ID:   24db56cb-5593-48b8-920b-b2a7c07176d4
```

The error disappears if I set  `download_only = "true"`: a local copy of already removed files in the shared library is preserved, and no upload attempt is registered as expected.

**Thus, a question arises here: how could I maintain a local copy mirroring read-only shared libraries? Or if  `download_only = "true"`, how could I get rid of local files no longer existing on the online account?**

Besides, I don't get a record with the file "in troubles" in the log file. Here are the records for the last two restarts of the concerned service with `enable_logging = "true"`:

```
2022-Aug-12 09:56:51.9935897    Starting a sync with OneDrive
2022-Aug-12 09:57:07.9949625
ERROR: Microsoft OneDrive API returned an error with the following message:
2022-Aug-12 09:57:07.9951661      Error Message:    HTTP request returned status code 403 (Forbidden)
2022-Aug-12 09:57:07.9952467      Error Reason:     Access denied
2022-Aug-12 09:57:07.9952838      Error Timestamp:  2022-08-12T07:57:07
2022-Aug-12 09:57:07.9953171      API Request ID:   610e440e-c4ed-41b9-9439-e94942b523f7
2022-Aug-12 09:57:11.9287477    Sync with OneDrive is complete
2022-Aug-12 10:02:11.9846092    Starting a sync with OneDrive
2022-Aug-12 10:02:28.1934432
ERROR: Microsoft OneDrive API returned an error with the following message:
2022-Aug-12 10:02:28.1935148      Error Message:    HTTP request returned status code 403 (Forbidden)
2022-Aug-12 10:02:28.1935941      Error Reason:     Access denied
2022-Aug-12 10:02:28.1936318      Error Timestamp:  2022-08-12T08:02:28
2022-Aug-12 10:02:28.1936655      API Request ID:   13e4e3cb-d0ae-41de-a002-479d5ed87aec
2022-Aug-12 10:02:32.1292428    Sync with OneDrive is complete
```

Please, let me know if I must create a separate discussion for the query about logging. I'm adding it here as it is much easier to establish the relationship between both issues. Thanks!

**Comment by abraunegg** [comment]
@rrodrigueznt 
> tries to upload its local copy as it detects it is lacking in the server side, and it fails as it has read-only access.

100% correct and the error message generated is correct - you cant upload when you have read-only permissions on OneDrive.

> The error disappears if I set `download_only = "true"`

100% correct as the application does not perform the upload sequencing when this flag is set, thus, this is why you do not get the error message that it cant upload the file.

> Thus, a question arises here: how could I maintain a local copy mirroring read-only shared libraries? Or if download_only = "true", how could I get rid of local files no longer existing on the online account?

If the file | folder was already in-sync, when it is deleted online, it will be deleted locally. This is confirmed functionality and works as advertised.

If the file | folder was *not* in-sync or you use the `--resync` option, the client has no idea what to do about this extra file locally, so it preserves the additional data.

It sounds like what you are asking for is a feature request so that local data can be cleaned up automatically regardless of if the sync state is known or not when using `--download-only`. Please raise a feature request.

> Please, let me know if I must create a separate discussion for the query about logging

This appears to be a bug ... which is odd ... as the function to write the logs out is the same. Please raise a bug report following the correct process: https://github.com/abraunegg/onedrive#reporting-an-issue-or-bug

**Update:**
* Created Feature Request: https://github.com/abraunegg/onedrive/issues/2092
* Created Bug Report: https://github.com/abraunegg/onedrive/issues/2093

**Comment by abraunegg** [comment]
@rrodrigueznt 
> The error disappears if I set `download_only = "true"`

Added further clarifications as per here: https://github.com/abraunegg/onedrive/blob/master/docs/application-security.md#configuring-read-only-access-to-your-onedrive-data

**Reply by rrodrigueznt** [reply]
Following the new contents about read-only access in a `CentOS Stream 8` server running `onedrive v2.4.20`, I get:
```
(base) [rjr@e220i059h057w onedrive_igfae-dir]$ onedrive --monitor --confdir=/home/rjr/.config/onedrive_igfae-dir --reauth
Unknown key in config file: read_only_auth_scope
Configuration file has errors - please check your configuration
(base) [rjr@e220i059h057w onedrive_igfae-dir]$ onedrive --monitor --confdir=/home/rjr/.config/onedrive_igfae-dir
Unknown key in config file: read_only_auth_scope
Configuration file has errors - please check your configuration
(base) [rjr@e220i059h057w onedrive_igfae-dir]$
```
Here the last few lines of the concerned config file:
```
# webhook_expiration_interval = "86400"
# webhook_renewal_interval = "43200"
# space_reservation = "50"
# display_running_config = "false"
read_only_auth_scope = "true"
```
Please, where am I wrong? Thanks!

**Reply by abraunegg** [reply]
@rrodrigueznt 
That value was introduced post v2.4.20 release - it is not valid for v2.4.20.

See: https://github.com/abraunegg/onedrive/pull/2065
